### PR TITLE
Refactor refprog for virtual hubs

### PR DESCRIPTION
**Pull request by alexandrem-rte**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceExchangeData.java [hunk 1]
```diff
@@ -6,52 +6,54 @@
...
 
-    boolean isAreaOutToAreaInExchange(Country areaOutId, Country areaInId) {
-        return this.areaIn != null && this.areaOut != null && this.areaIn.equals(areaInId) && this.areaOut.equals(areaOutId);
+    boolean isAreaOutToAreaInExchange(ReferenceProgramArea areaOut, ReferenceProgramArea areaIn) {
+        return this.areaIn != null && this.areaOut != null && this.areaIn.equals(areaIn) && this.areaOut.equals(areaOut);
+    }
+
+    public boolean isVirtualHub() {
+        return areaIn == null || areaIn.isVirtualHub() || areaOut == null || areaOut.isVirtualHub();
     }
```
Comment (baptsegu) [inline data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceExchangeData.java:58]: I find this method quite confusing : 
* a ReferenceExchangeData involves two areas, one can be a virtual hub but not the other ... so the name of the method seems a bit strange to me 
* when we build the ReferenceProgram with the `CountryNetPositionComputation` class, we directly computes the net-position of countries (and not the country-to-country exchanges), and therefore build ReferenceExchangeData with {country1="BE", country2=null} ... So I am not sure the `areaOut == null` works as expected in all cases

I would prefer to delete it.

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java [hunk 2]
```diff
@@ -19,56 +16,60 @@
...
     public ReferenceProgram(List<ReferenceExchangeData> referenceExchangeDataList) {
         this.referenceExchangeDataList = referenceExchangeDataList;
-        this.countries = new HashSet<>();
+        this.referenceProgramAreas = new HashSet<>();
         this.referenceExchangeDataList.stream().forEach(referenceExchangeData -> {
-            if (referenceExchangeData.getAreaOut() != null) {
-                countries.add(referenceExchangeData.getAreaOut());
+            referenceProgramAreas.add(referenceExchangeData.getAreaOut());
+            referenceProgramAreas.add(referenceExchangeData.getAreaIn());
+            if (referenceExchangeData.getAreaOut() != null && !referenceExchangeData.getAreaOut().isVirtualHub()) {
```
Comment (baptsegu) [inline data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java:28]: why do you filter the referenceExchangeData for which the areaOut is virtual ?
Will it be changed in the PR on the loop-flow computation ?

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java [hunk 3]
```diff
@@ -19,56 +16,60 @@
...
+                referenceProgramAreas.add(referenceExchangeData.getAreaOut());
             }
-            if (referenceExchangeData.getAreaIn() != null) {
-                countries.add(referenceExchangeData.getAreaIn());
+            if (referenceExchangeData.getAreaIn() != null && !referenceExchangeData.getAreaIn().isVirtualHub()) {
```
Comment (baptsegu) [inline data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java:31]: same

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java [hunk 4]
```diff
@@ -47,23 +47,29 @@ public static ReferenceProgram importRefProg(InputStream inputStream, OffsetDate
...
-            Country outArea = null;
-            Country inArea = null;
+            ReferenceProgramArea outArea;
+            String outAreaValue = timeSeries.getOutArea().getV();
             try {
-                outArea = (new EICode(timeSeries.getOutArea().getV())).getCountry();
+                new EICode(outAreaValue).getCountry();
+                outArea = new ReferenceProgramArea(outAreaValue);
             } catch (IllegalArgumentException e) {
                 LOGGER.warn("EIC code {} is not mapped to a country. The flow from this area will not be saved.", timeSeries.getOutArea().getV());
```
Comment (baptsegu) [inline data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java:56]: same as previous remark :p

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java [hunk 5]
```diff
@@ -47,23 +47,29 @@ public static ReferenceProgram importRefProg(InputStream inputStream, OffsetDate
...
+                outArea = new ReferenceProgramArea(timeSeries.getOutArea().getV());
             }
+            ReferenceProgramArea inArea;
+            String inAreaValue = timeSeries.getInArea().getV();
             try {
-                inArea = (new EICode(timeSeries.getInArea().getV())).getCountry();
+                new EICode(inAreaValue).getCountry();
+                inArea = new ReferenceProgramArea(inAreaValue);
             } catch (IllegalArgumentException e) {
                 LOGGER.warn("EIC code {} is not mapped to a country. The flow to this area will not be saved.", timeSeries.getInArea().getV());
```
Comment (baptsegu) [inline data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java:65]: same

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java [hunk 6]
```diff
@@ -47,23 +47,29 @@ public static ReferenceProgram importRefProg(InputStream inputStream, OffsetDate
...
+                inArea = new ReferenceProgramArea(timeSeries.getInArea().getV());
             }
-            if (outArea != null || inArea != null) {
+            if (inArea.isVirtualHub() && outArea.isVirtualHub()) {
+                LOGGER.warn("Neither origin ({}) nor extremity ({}) EIC code is mapped to a country. The flow will not be imported.", outAreaValue, inAreaValue);
```
Comment (baptsegu) [inline data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java:69]: I think we should import flows between virtual hubs, it make sense from a functional viewpoint, and I would not be surprise that we have some of these in HVDC modelling.

Cannot we keep the previous if :              `if (outArea != null || inArea != null) {`  ?

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java [hunk 7]
```diff
@@ -19,56 +16,60 @@
...
+        referenceProgramAreas.forEach(country -> netPositions.put(country, computeGlobalNetPosition(country)));
     }
 
     public List<ReferenceExchangeData> getReferenceExchangeDataList() {
         return referenceExchangeDataList;
     }
 
-    public Set<Country> getListOfCountries() {
-        return countries;
+    public Set<ReferenceProgramArea> getListOfCountries() {
```
Comment (baptsegu) [inline data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java:43]: Could you rename this method : getListOfAreas() ?

**Review by baptsegu — CHANGES_REQUESTED** [review]
Some questions and changes in the review.

One bigger suggestion on the name of classes, shouldn't we rename the classes :

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceExchangeData.java (hunk 1 above)
Comment (alexandrem-rte) [reply data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceExchangeData.java]: I have renamed this method "involvesVirtualHub". It is needed in LoopFlowComputation for the moment, I think we'll be able to delete it once the branch of farao-virtual-hubs is released.

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java (hunk 2 above)
Comment (alexandrem-rte) [reply data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java]: You're right, I removed some useless lines here (areaOut and areaIn were added in all cases so these checks were useless)

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java (hunk 3 above)
Comment (alexandrem-rte) [reply data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java]: same

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java (hunk 4 above)
Comment (alexandrem-rte) [reply data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java]: Indeed, I have removed this useless try/catch block

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java (hunk 5 above)
Comment (alexandrem-rte) [reply data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java]: same

File: data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java (hunk 6 above)
Comment (alexandrem-rte) [reply data/refprog/refprog-xml-importer/src/main/java/com/farao_community/farao/data/refprog/refprog_xml_importer/RefProgImporter.java]: ok, I was not sure if it made sense. I removed this and adapted a unit test accordingly.

File: data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java (hunk 7 above)
Comment (alexandrem-rte) [reply data/refprog/reference-program/src/main/java/com/farao_community/farao/data/refprog/reference_program/ReferenceProgram.java]: sure!

**Review by baptsegu — APPROVED** [review]
ok for me :)

**Pull request merged**